atl-checker/src/algorithms/global/com.rs

```diff
@@ -0,0 +1,164 @@
+    Result { task: V, value: bool },
+    Terminate,
+}
+
+pub trait GBroker<V: Hash + Eq + PartialEq + Clone> {
+    fn send_terminate(&self);
+    fn send_result(&self, task: V, value: bool);
+    fn receive(&self) -> Result<Option<GMessage<V>>, Box<dyn Error>>;
+    fn get_task(&self) -> Result<Option<V>, Box<dyn Error>>;
+    fn task_queue_is_empty(&self) -> bool;
+}
+
+pub trait GBrokerManager<V: Hash + Eq + PartialEq + Clone> {
+    fn send_updates(&self, assignments: HashMap<V, bool>);
+    fn queue_task(&self, task: V);
+    fn receive(&self) -> Result<GMessage<V>, Box<dyn Error>>;
+    fn task_queue_is_empty(&self) -> bool;
+    fn terminate(&self);
+}
+
+pub struct GChannelBrokerManager<V: Hash + Eq + PartialEq + Clone> {
+    workers: Vec<Sender<GMessage<V>>>,
+    updates: Receiver<GMessage<V>>,
+    task_queue_producer: Sender<V>,
+    task_queue_consumer: Receiver<V>,
+}
+
+impl<V: Hash + Eq + PartialEq + Clone> GBrokerManager<V> for GChannelBrokerManager<V> {
+    fn send_updates(&self, updates: HashMap<V, bool>) {
+        let msg = GMessage::Updates { updates };
+        for worker in &self.workers {
+            worker
+                .send(msg.clone())
+                .expect("Sending assignments to workers failed");
+        }
+    }
+    fn queue_task(&self, task: V) {
+        self.task_queue_producer
+            .send(task)
+            .expect("Sending tasks failed");
+    }
+
+    fn receive(&self) -> Result<GMessage<V>, Box<dyn Error>> {
+        match self.updates.try_recv() {
+            Ok(msg) => Ok(msg),
+            Err(err) => match err {
+                TryRecvError::Empty => {
+                    debug!("nothing to receive");
+                    Ok(GMessage::Updates {
+                        updates: HashMap::new(),
+                    })
+                }
+                TryRecvError::Disconnected => Err(Box::new(err)),
+            },
+        }
+    }
+
+    fn task_queue_is_empty(&self) -> bool {
+        self.task_queue_consumer.is_empty()
+    }
+
+    fn terminate(&self) {
+        for to in 0..self.workers.len() {
+            let _err = self
+                .workers
+                .get(to as usize)
+                .expect("receiver id out of bounds")
+                .send(GMessage::Terminate);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct GChannelBroker<V: Hash + Eq + PartialEq + Clone> {
+    updates: Sender<GMessage<V>>,
+    receiver: Receiver<GMessage<V>>,
+    task_queue_consumer: Receiver<V>,
+}
+
+impl<V: Hash + Eq + PartialEq + Clone> GBroker<V> for GChannelBroker<V> {
+    fn send_terminate(&self) {
+        self.updates
+            .send(GMessage::Terminate)
+            .expect("Failed to send terminate message");
+    }
+
+    fn send_result(&self, task: V, value: bool) {
+        self.updates
+            .send(GMessage::Result { task, value })
+            .expect("Failed to send the assignments table to main thread");
+    }
```

Comment:
I spot some copy-pasting for the local algorithm. Double check that the function names and error messages still make sense. For instance, here is says "assignments table" even though it is only a single assignment.